File: src/core/Stocks/PendingStockPositionEvents.cs
using System;
using core.Shared;

namespace core.Stocks
{
    internal class PendingStockPositionCreated : AggregateEvent
    {
        public PendingStockPositionCreated(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string ticker,
            decimal price,
            decimal numberOfShares,
            decimal? stopPrice,
            string notes)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Ticker = ticker;
            Price = price;
            NumberOfShares = numberOfShares;
            StopPrice = stopPrice;
            Notes = notes;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal Price { get; }
        public decimal NumberOfShares { get; }
        public decimal? StopPrice { get; }
        public string Notes { get; }
    }
    
    internal class PendingStockPositionCreatedWithStrategy : AggregateEvent
    {
        public PendingStockPositionCreatedWithStrategy(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            string notes,
            decimal numberOfShares,
            decimal price,
            decimal? stopPrice,
            string strategy,
            string ticker,
            Guid userId)
            : base(id, aggregateId, when)
        {
            Notes = notes;
            NumberOfShares = numberOfShares;
            Price = price;
            StopPrice = stopPrice;
            Strategy = strategy;
            Ticker = ticker;
            UserId = userId;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal Price { get; }
        public decimal NumberOfShares { get; }
        public decimal? StopPrice { get; }
        public string Notes { get; }
        public string Strategy { get; }
    }
    
    internal class PendingStockPositionCreatedWithStrategyAndSizeStop : AggregateEvent
    {
        public PendingStockPositionCreatedWithStrategyAndSizeStop(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            string notes,
            decimal numberOfShares,
            decimal price,
            decimal? stopPrice,
            decimal? sizeStopPrice,
            string strategy,
            string ticker,
            Guid userId)
            : base(id, aggregateId, when)
        {
            Notes = notes;
            NumberOfShares = numberOfShares;
            Price = price;
            StopPrice = stopPrice;
            SizeStopPrice = sizeStopPrice;
            Strategy = strategy;
            Ticker = ticker;
            UserId = userId;
        }

        public Guid UserId { get; }
        public string Ticker { get; }
        public decimal Price { get; }
        public decimal NumberOfShares { get; }
        public decimal? StopPrice { get; }
        public decimal? SizeStopPrice { get; }
        public string Notes { get; }
        public string Strategy { get; }
    }
    
    public class PendingStockPositionRealized : AggregateEvent
    {
        public PendingStockPositionRealized(Guid id, Guid aggregateId, DateTimeOffset when, decimal price)
            : base(id, aggregateId, when)
        {
            Price = price;
        }

        public decimal Price { get; }
    }

    public class PendingStockPositionOrderDetailsAdded : AggregateEvent
    {
        public PendingStockPositionOrderDetailsAdded(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            string orderType,
            string orderDuration)
            : base(id, aggregateId, when)
        {
            OrderType = orderType;
            OrderDuration = orderDuration;
        }
        
        public string OrderType { get; }
        public string OrderDuration { get; }
    }

    public class PendingStockPositionClosed : AggregateEvent
    {
        public PendingStockPositionClosed(Guid id, Guid aggregateId, DateTimeOffset when, bool purchased, decimal? price)
            : base(id, aggregateId, when)
        {
            Purchased = purchased;
            Price = price;
        }

        public bool Purchased { get; }
        public decimal? Price { get; }
    }
    
    public class PendingStockPositionClosedWithReason : AggregateEvent
    {
        public PendingStockPositionClosedWithReason(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            string reason)
            : base(id, aggregateId, when)
        {
            Reason = reason;
        }

        public string Reason { get; }
    }
}

File: src/core2.fs/Stocks/PendingStockPositionEvents.fs
namespace core.Stocks

open System
open core.Shared

[<AllowNullLiteral>]
type internal PendingStockPositionCreated(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, ticker: string, price: decimal, numberOfShares: decimal, stopPrice: decimal option, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val Price = price with get
    member val NumberOfShares = numberOfShares with get
    member val StopPrice = stopPrice with get
    member val Notes = notes with get

[<AllowNullLiteral>]
type internal PendingStockPositionCreatedWithStrategy(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, notes: string, numberOfShares: decimal, price: decimal, stopPrice: decimal option, strategy: string, ticker: string, userId: Guid) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val Price = price with get
    member val NumberOfShares = numberOfShares with get
    member val StopPrice = stopPrice with get
    member val Notes = notes with get
    member val Strategy = strategy with get

[<AllowNullLiteral>]
type internal PendingStockPositionCreatedWithStrategyAndSizeStop(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, notes: string, numberOfShares: decimal, price: decimal, stopPrice: decimal option, sizeStopPrice: decimal option, strategy: string, ticker: string, userId: Guid) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Ticker = ticker with get
    member val Price = price with get
    member val NumberOfShares = numberOfShares with get
    member val StopPrice = stopPrice with get
    member val SizeStopPrice = sizeStopPrice with get
    member val Notes = notes with get
    member val Strategy = strategy with get

[<AllowNullLiteral>]
type PendingStockPositionRealized(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, price: decimal) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Price = price with get

[<AllowNullLiteral>]
type PendingStockPositionOrderDetailsAdded(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, orderType: string, orderDuration: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val OrderType = orderType with get
    member val OrderDuration = orderDuration with get

[<AllowNullLiteral>]
type PendingStockPositionClosed(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, purchased: bool, price: decimal option) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Purchased = purchased with get
    member val Price = price with get

[<AllowNullLiteral>]
type PendingStockPositionClosedWithReason(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, reason: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Reason = reason with get
